### PR TITLE
Added support for unix domain sockets to the sockets transport

### LINKED ISSUE #4764: Allow Unix Sockets on Windows
😮 https://blogs.msdn.microsoft.com/commandline/2017/12/19/af_unix-comes-to-windows/

### LINKED ISSUE #4768: Consider supporting unix domain sockets with SocketTransport
Seems like it's possible. 

https://github.com/dotnet/corefx/blob/958511d023231f120109967484fb63ae1837e6b8/src/System.IO.Pipes/src/System/Net/Sockets/UnixDomainSocketEndPoint.cs

https://github.com/dotnet/corefx/blob/b5d0c3277dc384a23fee2e57a0b5c232c99dbd88/src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Unix.cs#L290-L301

It would be cleaner if we kept using the socket APIs and could just copy that end point logic over to kestrel. Ideally we wouldn't have to use the pipe API to support this.'

This might be the only problem https://github.com/dotnet/corefx/blob/958511d023231f120109967484fb63ae1837e6b8/src/System.IO.Pipes/src/System/Net/Sockets/UnixDomainSocketEndPoint.cs#L25

/cc @stephentoub

### PR DESCRIPTION
**Pull request by davidfowl**
- Added a test for UnixDomainSockets on both transports

Fixes #4768
Fixes #4764

PS: This is on top of #10321. This will be rebased once that's in.

### CODE REVIEW

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs [hunk 1]
```diff
@@ -66,9 +62,16 @@ internal void Bind()
                 throw new InvalidOperationException(SocketsStrings.TransportAlreadyBound);
             }
 
-            // TODO: Add support for UnixDomainSocket
+            Socket listenSocket;
```
Comment (gfoidl) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs]: Maybe like this
```c#
Socket listenSocket = new Socket(EndPoint.AddressFamily, SocketType.Stream, EndPoint is UnixDomainSocketEndPoint ? ProtocolType.Unspecified : ProtocolType.Tcp);
```
?

Comment (davidfowl) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs]: Yep, I was thinking about that after I saw this PR 😄

**Comment by anurse** [comment]
@davidfowl Do you expect to be able to get this in and verified e2e by Friday? Assuming not, we should wait until preview 7 (opens next week) to merge.

**Comment by davidfowl** [comment]
> @davidfowl Do you expect to be able to get this in and verified e2e by Friday? Assuming not, we should wait until preview 7 (opens next week) to merge.

Yes, I expect it to be in today or tomorrow.

File: src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                var hostBuilder = TransportSelector.GetWebHostBuilder()
+                    .UseKestrel(o =>
+                    {
+                        o.ListenUnixSocket(path, builder =>
+                        {
+                            builder.Run(EchoServer);
+                        });
+                    })
+                    .ConfigureServices(AddTestLogging)
+                    .Configure(c => { });
```
Comment (halter73) [inline src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs]: Out of curiosity, is this line necessary?

File: src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                        socket.Connect(new UnixDomainSocketEndPoint(path));
+
+                        var data = Encoding.ASCII.GetBytes("Hello World");
+                        socket.Send(data);
+
+                        var buffer = new byte[data.Length];
+                        var read = 0;
+                        while (read < data.Length)
+                        {
+                            read += socket.Receive(buffer, read, buffer.Length - read, SocketFlags.None);
```
Comment (halter73) [inline src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs]: I know it's just a test, but I'd avoid blocking. We've seen threadpool issues in the past...

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs [hunk 4]
```diff
@@ -31,10 +31,6 @@ internal sealed class SocketConnectionListener : IConnectionListener
             SocketTransportOptions options,
             ISocketsTrace trace)
         {
-            Debug.Assert(endpoint != null);
```
Comment (jkotalik) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs]: `endpoint != null` is still a valid assert I believe.

File: src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs (hunk 2 above)
Comment (davidfowl) [reply src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs]: Yea, the webhost throws without it.

File: src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    {
+#if LIBUV
+        [OSSkipCondition(OperatingSystems.Windows, SkipReason = "Libuv does not support unix domain sockets on Windows.")]
+#else
+        [OSSkipCondition(OperatingSystems.Windows, WindowsVersions.Win7, WindowsVersions.Win8, WindowsVersions.Win81, WindowsVersions.Win2008R2, SkipReason = "UnixDomainSocketEndPoint is not supported on older versions of Windows")]
+#endif
+        [ConditionalFact]
+        public async Task TestUnixDomainSocket()
+        {
+            var path = Path.GetTempFileName();
```
Comment (halter73) [inline src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs]: This seems to be creating the file.

**Comment by davidfowl** [comment]
@halter73 @jkotalik is this good to go?

File: src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs (hunk 5 above)
Comment (jkotalik) [reply src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs]: Yeah this is a bit jank.

**Review by jkotalik — APPROVED** [review]

File: src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs (hunk 5 above)
Comment (anurse) [reply src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs]: Especially when you're using it just to get a name and then immediately deleting it. This is a bit of a flake-farm waiting to happen... why not just use `Path.Combine(Path.GetTempPath(), $"{Guid.NewGuid():N}.socket")` or something?

**Pull request merged**